File: solutions/ruby/minesweeper/minesweeper.rb
class Minesweeper
  def self.annotate(*args) = new(*args).annotate

  def initialize(input)
    @minefield = input.map(&:chars)
    @rows = @minefield.length
    @cols = @minefield[0]&.length.to_i
  end

  def annotate
    minefield.each_with_index do |row, r|
      row.each_with_index do |cell, c|
        increment_neighbors(r, c) if cell == "*"
      end
    end
    minefield.map(&:join)
  end

  private

  attr_reader :minefield, :rows, :cols

  def increment_neighbors(row, col)
    [-1, 0, 1].each do |row_offset|
      [-1, 0, 1].each do |col_offset|
        increment_neighbor(row + row_offset, col + col_offset)
      end
    end
  end

  def increment_neighbor(row, col)
    return if row < 0 || row > @rows - 1
    return if col < 0 || col > @cols - 1

    value = minefield[row][col]
    return if value == "*"

    minefield[row][col] = value.to_i + 1
  end
end

File: solutions/javascript/minesweeper/minesweeper.js
//
// Add the mine counts to empty squares in a completed Minesweeper board.
//
const NEIGHBORS = [
  [-1, -1],
  [-1, 0],
  [-1, 1],
  [0, -1],
  [0, 1],
  [1, -1],
  [1, 0],
  [1, 1],
];
const MINE = "*";

export const annotate = (input) => {
  // check for no rows or columns
  if (!input.length || input[0] === "") return input;

  // create matrix of minefield as 2D array
  let minefield = input.map((str) => [...str]);
  const rows = minefield.length;
  const cols = minefield[0].length;

  const incrementNeighbor = (row, col) => {
    // check neighbor is in the minefield
    if (row < 0 || row >= rows) return;
    if (col < 0 || col >= cols) return;

    // check neighbor is not another mine
    const value = minefield[row][col];
    if (value === MINE) return;

    // increment the count for this cell
    minefield[row][col] = Number(value) + 1;
  };

  const incrementNeighbors = (row, col) => {
    NEIGHBORS.forEach(([rowOffset, colOffset]) => {
      incrementNeighbor(row + rowOffset, col + colOffset);
    });
  };

  // Traverse minefield to find mine and increment cells around it
  minefield.forEach((row, r) => {
    row.forEach((cell, c) => {
      if (cell === MINE) incrementNeighbors(r, c);
    });
  });

  return minefield.map((row) => row.join(""));
};
